Sources/SwifterSwift/SwiftStdlib/ArrayExtensions.swift

```diff
@@ -97,6 +97,56 @@ public extension Array {
         return self
     }
 
+    /// SwifterSwift: Remove all duplicate elements from Array.
+    ///
+    ///        [1, -1, 2, -4, 3, 3, 3, 4, -5].removeDuplicates { abs($0) == abs($1) } -> [1, 2, -4, 3, -5]
+    ///
+    /// - Parameters:
+    ///   - comparator: Returns true if 2 elements of array are equal.
+    /// - Returns: Return array with all duplicate elements removed.
+    @discardableResult
+    mutating func removeDuplicates(by comparator: (Element, Element) -> Bool) -> [Element] {
+        self = withoutDuplicates(by: comparator)
+        return self
+    }
+
+    /// SwifterSwift: Remove all duplicate elements from Array.
+    ///
+    /// - Parameters:
+    ///   - path: Key path to compare, must be Equatable
+    /// - Returns: Return array with all duplicate elements removed.
+    @discardableResult
+    mutating func removeDuplicates<E: Equatable>(keyPath path: KeyPath<Element, E>) -> [Element] {
+        self = withoutDuplicates(keyPath: path)
+        return self
+    }
+
+    /// SwifterSwift: Return array with all duplicate elements removed.
+    ///
+    ///     [1, -1, 2, -4, 3, 3, 3, 4, -5].withoutDuplicates { abs($0) == abs($1) } -> [1, 2, -4, 3, -5]
+    ///
+    /// - Parameters:
+    ///   - comparator: Returns true if 2 elements of array are equal.
+    /// - Returns: an array of unique elements.
+    ///
+    func withoutDuplicates(by comparator: (Element, Element) -> Bool) -> [Element] {
+        // Thanks to https://github.com/sairamkotha for improving the method
+        return reduce(into: [Element]()) { result, element in
```

Comment:
We could provide a custom more performant version of this for Hashable?

Reply:
Hello @LucianoPAlmeida 
Do you mean [the version](https://github.com/SwifterSwift/SwifterSwift/issues/653#issuecomment-479406641) provided by @guykogus?

Reply:
@LucianoPAlmeida Also, the implementation based on `Set` is changing elements ordering. I am not sure if all programmers expect such behaviour.

Reply:
Kinda... I was thinking in something like:
```swift
func withoutDuplicates(by comparator: (Element, Element) -> Bool) -> [Element] where Element: Hashable {
    var set: Set<Element> = []
    return filter { set.insert($0) } 
}
// And for the mutating
mutating func withoutDuplicates(by comparator: (Element, Element) -> Bool) -> [Element] where Element: Hashable {
    var set: Set<Element> = []
    removeAll(where: { set.insert($0) }) // Since it makes a partition based removal is more optimized.
}
```
I think maybe this is valid because most usages will be with types like Int, String, double ... which are normally hashable. 
What do you think? @guykogus

Reply:
@LucianoPAlmeida But you don't call `comparator` in these functions. Could you please provide a version with this function?

Reply:
Ohh, forget about it I was thinking about the one with no params. Not sure if there is a good way to do it with a custom comparator closure :))